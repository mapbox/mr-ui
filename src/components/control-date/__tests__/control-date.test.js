import React from 'react';
import { shallow } from 'enzyme';
import { testCases } from './control-date-test-cases';
import moment from 'moment';

function getDatePickerProps(wrapper) {
  return wrapper.find('DatePicker').first().props();
}
function getEndDatePickerProps(wrapper) {
  return wrapper.find('DatePicker').last().props();
}

describe('ControlDate', () => {
  let testCase;
  let wrapper;
  let datePickerProps;
  let startDatePickerProps;
  let endDatePickerProps;

  describe(testCases.single.description, () => {
    beforeEach(() => {
      testCase = testCases.single;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
      datePickerProps = getDatePickerProps(wrapper);
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('updates from calendar selection', () => {
      const pickedDate = moment();

      datePickerProps.onChange(pickedDate);
      wrapper.update();
      datePickerProps = getDatePickerProps(wrapper);
      expect(datePickerProps.selected).toBe(pickedDate);
      expect(testCase.props.onChange).toHaveBeenCalledTimes(1);
      expect(testCase.props.onChange).toHaveBeenCalledWith(
        { date: pickedDate.toDate() },
        testCase.props.id
      );
    });

    test('native datepicker selection works', () => {
      const nativeDate = '2017-08-09';
      const mockEvent = {
        target: {
          value: nativeDate
        }
      };
      const nativeDateMoment = moment(nativeDate, 'YYYY-MM-DD');

      const nativeDatepickerProps = wrapper
        .find('input[pattern]')
        .first()
        .props();
      nativeDatepickerProps.onChange(mockEvent);
      wrapper.update();
      datePickerProps = getDatePickerProps(wrapper);
      expect(datePickerProps.selected.toString()).toBe(
        nativeDateMoment.toString()
      );
      expect(testCase.props.onChange).toHaveBeenCalledTimes(1);
      expect(testCase.props.onChange).toHaveBeenCalledWith(
        { date: nativeDateMoment.toDate() },
        testCase.props.id
      );
    });
  });

  describe(testCases.double.description, () => {
    beforeEach(() => {
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
      testCase = testCases.double;
      startDatePickerProps = getDatePickerProps(wrapper);
      endDatePickerProps = getEndDatePickerProps(wrapper);
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('updates from calendar selection', () => {
      const pickedStartDate = moment();
      const pickedEndDate = moment().add(7, 'days');

      startDatePickerProps.onChange(pickedStartDate);
      wrapper.update();
      startDatePickerProps = getDatePickerProps(wrapper);
      expect(startDatePickerProps.selected).toBe(pickedStartDate);
      expect(testCase.props.onChange).toHaveBeenCalledTimes(1);
      expect(testCase.props.onChange).toHaveBeenCalledWith(
        {
          date: pickedStartDate.toDate(),
          endDate: testCase.props.value.endDate
        },
        testCase.props.id
      );

      endDatePickerProps.onChange(pickedEndDate);
      wrapper.update();
      endDatePickerProps = getEndDatePickerProps(wrapper);
      expect(endDatePickerProps.selected).toBe(pickedEndDate);
      expect(testCase.props.onChange).toHaveBeenCalledTimes(2);
      expect(testCase.props.onChange).toHaveBeenCalledWith(
        {
          date: pickedStartDate.toDate(),
          endDate: pickedEndDate.toDate()
        },
        testCase.props.id
      );
    });

    test('native datepicker selection works', () => {
      const nativeStartDate = '2017-08-09';
      const nativeEndDate = '2017-11-09';
      const mockEventOne = {
        target: {
          value: nativeStartDate
        }
      };
      const mockEventTwo = {
        target: {
          value: nativeEndDate
        }
      };
      const nativeStartDateMoment = moment(nativeStartDate, 'YYYY-MM-DD');
      const nativeEndDateMoment = moment(nativeEndDate, 'YYYY-MM-DD');

      const nativeStartDatepickerProps = wrapper
        .find('input[pattern]')
        .first()
        .props();
      const nativeEndDatepickerProps = wrapper
        .find('input[pattern]')
        .last()
        .props();

      nativeStartDatepickerProps.onChange(mockEventOne);
      wrapper.update();
      datePickerProps = getDatePickerProps(wrapper);
      expect(datePickerProps.selected.toString()).toBe(
        nativeStartDateMoment.toString()
      );
      expect(testCase.props.onChange).toHaveBeenCalledTimes(1);
      expect(testCase.props.onChange).toHaveBeenCalledWith(
        {
          date: nativeStartDateMoment.toDate(),
          endDate: testCase.props.value.endDate
        },
        testCase.props.id
      );

      nativeEndDatepickerProps.onChange(mockEventTwo);
      wrapper.update();
      endDatePickerProps = getEndDatePickerProps(wrapper);
      expect(endDatePickerProps.selected.toString()).toBe(
        nativeEndDateMoment.toString()
      );
      expect(testCase.props.onChange).toHaveBeenCalledTimes(2);
      expect(testCase.props.onChange).toHaveBeenCalledWith(
        {
          date: nativeStartDateMoment.toDate(),
          endDate: nativeEndDateMoment.toDate()
        },
        testCase.props.id
      );
    });
  });

  describe(testCases.disabled.description, () => {
    beforeEach(() => {
      testCase = testCases.disabled;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
      datePickerProps = getDatePickerProps(wrapper);
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('disabled prop works', () => {
      wrapper.find('input').forEach((input) => {
        expect(input.props()).toHaveProperty('disabled', true);
      });
    });
  });

  describe(testCases.restrictedRange.description, () => {
    beforeEach(() => {
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
      testCase = testCases.restrictedRange;
      startDatePickerProps = getDatePickerProps(wrapper);
      endDatePickerProps = getEndDatePickerProps(wrapper);
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('selecting date outside min/max range will fail', () => {
      const pickedStartDate = startDatePickerProps.minDate.subtract(1, 'days');
      const pickedEndDate = startDatePickerProps.minDate.add(1, 'days');

      startDatePickerProps.onChange(pickedStartDate);
      wrapper.update();
      expect(startDatePickerProps.selected).toBe(null);

      endDatePickerProps.onChange(pickedEndDate);
      wrapper.update();
      expect(endDatePickerProps.selected).toBe(null);
    });

    test('selecting date outside min/max range will fail', () => {
      const nativeStartDate = '2012-08-09';
      const nativeEndDate = '2020-11-09';
      const mockEventOne = {
        target: {
          value: nativeStartDate
        }
      };
      const mockEventTwo = {
        target: {
          value: nativeEndDate
        }
      };

      const nativeStartDatepickerProps = wrapper
        .find('input[pattern]')
        .first()
        .props();
      const nativeEndDatepickerProps = wrapper
        .find('input[pattern]')
        .last()
        .props();

      nativeStartDatepickerProps.onChange(mockEventOne);
      wrapper.update();
      datePickerProps = getDatePickerProps(wrapper);
      expect(datePickerProps.selected).toBe(null);

      nativeEndDatepickerProps.onChange(mockEventTwo);
      wrapper.update();
      endDatePickerProps = getEndDatePickerProps(wrapper);
      expect(endDatePickerProps.selected).toBe(null);
    });
  });

  describe(testCases.allOptions.description, () => {
    beforeEach(() => {
      testCase = testCases.allOptions;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );

      startDatePickerProps = getDatePickerProps(wrapper);
      endDatePickerProps = getEndDatePickerProps(wrapper);
    });

    test('renders', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('dateRange prop works', () => {
      const inputClassName = `.${testCase.props.themeTextInput
        .split(' ')
        .join('.')}`;
      expect(wrapper.find(inputClassName)).toHaveLength(2);
    });

    test('placeholder props work', () => {
      const inputClassName = `.${testCase.props.themeTextInput
        .split(' ')
        .join('.')}`;
      expect(wrapper.find(inputClassName).first().props().placeholderText).toBe(
        testCase.props.placeholder
      );
      expect(wrapper.find(inputClassName).last().props().placeholderText).toBe(
        testCase.props.endDatePlaceholder
      );
    });
  });
});
