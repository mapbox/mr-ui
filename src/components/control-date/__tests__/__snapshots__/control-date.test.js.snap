// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlDate All datepicker options renders 1`] = `
<ControlWrapper
  id="disabledDate"
  themeControlWrapper=""
>
  <div>
    <ControlLabel
      aside="A note on the date format"
      id="disabledDate"
      text="Disabled"
      themeLabel="txt-s txt-bold mb6"
    />
    <div
      className="react-control-date flex"
    >
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="border--0"
          className="input someFunSelector"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={false}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          id="disabledDate"
          maxDate={"2017-08-27T15:05:04.000Z"}
          minDate={"2017-08-12T15:05:04.000Z"}
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="Select a radical date"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={null}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={null}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          disabled={false}
          max="2017-08-27"
          min="2017-08-12"
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="Select a radical date"
          type="date"
          value=""
        />
      </div>
      <span
        className="px12"
      >
        to
      </span>
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="border--0"
          className="input someFunSelector"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={false}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          maxDate={"2017-08-27T15:05:04.000Z"}
          minDate={"2017-08-12T15:05:04.000Z"}
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="Select a radical end date"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={null}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={null}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          disabled={false}
          max="2017-08-27"
          min="2017-08-12"
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="Select a radical end date"
          type="date"
          value=""
        />
      </div>
    </div>
  </div>
</ControlWrapper>
`;

exports[`ControlDate Restricted Date range renders 1`] = `
<ControlWrapper
  id="disabledDate"
  themeControlWrapper=""
>
  <div>
    <ControlLabel
      id="disabledDate"
      text="Disabled"
      themeLabel="txt-s txt-bold mb6"
    />
    <div
      className="react-control-date react-control-date--disable-keyboard inline-flex flex--center-cross flex--wrap"
    >
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="shadow-darken10 border--0 none block-mm"
          className="input w180 none block-mm"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={true}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          id="disabledDate"
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="M/D/YYYY"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={null}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={null}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          disabled={true}
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="M/D/YYYY"
          type="date"
          value=""
        />
      </div>
      <span
        className="px12"
      >
        to
      </span>
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="shadow-darken10 border--0 none block-mm"
          className="input w180 none block-mm"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={true}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="M/D/YYYY"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={null}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={null}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          disabled={true}
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="M/D/YYYY"
          type="date"
          value=""
        />
      </div>
    </div>
  </div>
</ControlWrapper>
`;

exports[`ControlDate disabled datepicker renders 1`] = `
<ControlWrapper
  id="disabledDate"
  themeControlWrapper=""
>
  <div>
    <ControlLabel
      id="disabledDate"
      text="Disabled"
      themeLabel="txt-s txt-bold mb6"
    />
    <div
      className="react-control-date react-control-date--disable-keyboard inline-flex flex--center-cross flex--wrap"
    >
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="shadow-darken10 border--0 none block-mm"
          className="input w180 none block-mm"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={true}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          id="disabledDate"
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="M/D/YYYY"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={null}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={null}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          disabled={true}
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="M/D/YYYY"
          type="date"
          value=""
        />
      </div>
      <span
        className="px12"
      >
        to
      </span>
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="shadow-darken10 border--0 none block-mm"
          className="input w180 none block-mm"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={true}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="M/D/YYYY"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={null}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={null}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          disabled={true}
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="M/D/YYYY"
          type="date"
          value=""
        />
      </div>
    </div>
  </div>
</ControlWrapper>
`;

exports[`ControlDate double datepicker renders 1`] = `
<ControlWrapper
  id="singleDate"
  themeControlWrapper=""
>
  <div>
    <ControlLabel
      id="singleDate"
      text="Single Datepicker"
      themeLabel="txt-s txt-bold mb6"
    />
    <div
      className="react-control-date react-control-date--disable-keyboard inline-flex flex--center-cross flex--wrap"
    >
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="shadow-darken10 border--0 none block-mm"
          className="input w180 none block-mm"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={false}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          id="singleDate"
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="M/D/YYYY"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={"2017-08-15T15:05:04.000Z"}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={"2017-08-15T15:05:04.000Z"}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="M/D/YYYY"
          type="date"
          value="2017-08-15"
        />
      </div>
    </div>
  </div>
</ControlWrapper>
`;

exports[`ControlDate single datepicker renders 1`] = `
<ControlWrapper
  id="singleDate"
  themeControlWrapper=""
>
  <div>
    <ControlLabel
      id="singleDate"
      text="Single Datepicker"
      themeLabel="txt-s txt-bold mb6"
    />
    <div
      className="react-control-date react-control-date--disable-keyboard inline-flex flex--center-cross flex--wrap"
    >
      <div
        className="flex"
      >
        <DatePicker
          allowSameDay={false}
          calendarClassName="shadow-darken10 border--0 none block-mm"
          className="input w180 none block-mm"
          dateFormat="M/D/YYYY"
          dateFormatCalendar="MMMM YYYY"
          disabled={false}
          disabledKeyboardNavigation={true}
          dropdownMode="scroll"
          endDate={null}
          id="singleDate"
          monthsShown={1}
          onBlur={[Function]}
          onChange={[Function]}
          onChangeRaw={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          placeholderText="M/D/YYYY"
          popperModifiers={
            Object {
              "preventOverflow": Object {
                "boundariesElement": "viewport",
                "enabled": true,
                "escapeWithReference": true,
              },
            }
          }
          selected={"2017-08-15T15:05:04.000Z"}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          startDate={"2017-08-15T15:05:04.000Z"}
          timeIntervals={30}
          utcOffset={0}
          withPortal={false}
        />
        <input
          className="input input-s w180 block none-mm"
          onChange={[Function]}
          pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
          placeholder="M/D/YYYY"
          type="date"
          value="2017-08-15"
        />
      </div>
    </div>
  </div>
</ControlWrapper>
`;
